File: SpeedTestGameClient/core/src/main/java/si/um/feri/speedii/classes/MobileTower.java
package si.um.feri.speedii.classes;

import org.bson.types.ObjectId;

public class MobileTower {

    private Location location;
    private String provider;
    private String type;
    private boolean confirmed;
    private User locator;
    private final ObjectId id;

    public MobileTower(Location location, String provider, String type, boolean confirmed, User locator, ObjectId id) {
        this.location = location;
        this.provider = provider;
        this.type = type;
        this.confirmed = confirmed;
        this.locator = locator;
        this.id = id;
    }

    public MobileTower(Location location, String provider, String type, boolean confirmed, User locator) {
        this(location, provider, type, confirmed, locator, new ObjectId());
    }

    @Override
    public String toString() {
        return location + ", " + provider + ", " + type + ", " + confirmed + ", " + locator + " " + id;
    }

    // Getters and Setters
    public Location getLocation() {
        return location;
    }

    public void setLocation(Location location) {
        this.location = location;
    }

    public String getProvider() {
        return provider;
    }

    public void setProvider(String provider) {
        this.provider = provider;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public boolean isConfirmed() {
        return confirmed;
    }

    public void setConfirmed(boolean confirmed) {
        this.confirmed = confirmed;
    }

    public User getLocator() {
        return locator;
    }

    public void setLocator(User locator) {
        this.locator = locator;
    }

    public ObjectId getId() {
        return id;
    }
}

File: MobileApp/app/speedtest/src/main/java/sosteric/pora/speedtest/MobileTower.kt
import android.os.Parcel
import android.os.Parcelable
import org.bson.types.ObjectId

class MobileTower(
    var location: Location,
    var provider: String,
    var type: String,
    var confirmed: Boolean,
    var locator: User?,
    var id: ObjectId = ObjectId()
) : Parcelable {
    override fun toString(): String {
        return "${location}, $provider, $type, $confirmed, $locator $id"
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeParcelable(location, flags)
        parcel.writeString(provider)
        parcel.writeString(type)
        parcel.writeByte(if (confirmed) 1 else 0)
        parcel.writeParcelable(locator, flags)
        parcel.writeSerializable(id)
    }

    class MobileTowerAlt(
        var location: Location,
        var provider: String,
        var type: String,
        var confirmed: Boolean,
        var locator: String?,
        var id: ObjectId
    )

    fun toAlt(): MobileTowerAlt {
        val currentLocator = locator
        val locatorId = if(locator != null) { currentLocator!!.id.toString() } else { null }
        return MobileTowerAlt(location, provider, type, confirmed, locatorId, id)
    }
}